### PR TITLE
Replace uses of `MaybeUninit::uninit_array()` with inline const blocks.

### PR DESCRIPTION
**Pull request by kpreid**
\[This PR originally contained the changes in #125995 too. See edit history for the original PR description.]

The documentation of `MaybeUninit::uninit_array()` says:

> Note: in a future Rust version this method may become unnecessary when Rust allows [inline const expressions](https://github.com/rust-lang/rust/issues/76001). The example below could then use `let mut buf = [const { MaybeUninit::<u8>::uninit() }; 32];`.

The PR adding it also said: <https://github.com/rust-lang/rust/pull/65580#issuecomment-544200681>
    
> if it’s stabilized soon enough maybe it’s not worth having a standard library method that will be replaceable with `let buffer = [MaybeUninit::<T>::uninit(); $N];`

That time has come to pass — inline const expressions are stable — so `MaybeUninit::uninit_array()` is now unnecessary. The only remaining question is whether it is an important enough *convenience* to keep it around.

I believe it is net good to remove this function, on the principle that it is better to compose two orthogonal features (`MaybeUninit` and array construction) than to have a specific function for the specific combination, now that that is possible.

### CODE REVIEW

**Comment by ChayimFriedman2** [comment]
I still prefer `MaybeUninit::uninit_array()` to `[const { MaybeUninit::uninit() }; N]`, it's shorter and more readable IMHO.

**Comment by workingjubilee** [comment]
The examples and the compiler code should probably still prefer the stable option.

Technically the compared-against example was `[MaybeUninit::uninit(); N]`, which implies that the array initializer is a const context.

**Comment by Nilstrieb** [comment]
this fell down into the "i hope to get to this eventually but will probably never lol" part of my notification queue and i forgot to look at it again, sorry.
you have my approval for the first two commits, but I'd like a libs-api opinion on the last one. you can split out the first two into a new PR and i'll approve that, or you can keep them all like this if you're okay with letting the first two wait longer (as it will probably take a bit).
r? libs-api

**Comment by kpreid** [comment]
Split out the first two parts into #125995. I'll re-summarize this PR once that's merged.

**Comment by kpreid** [comment]
This PR now consists solely of removing `MaybeUninit::uninit_array()`.

File: compiler/rustc_data_structures/src/sip128.rs
Comment (kpreid) [reply compiler/rustc_data_structures/src/sip128.rs]: Fixed this and several others.

File: library/core/src/fmt/float.rs
Comment (kpreid) [reply library/core/src/fmt/float.rs]: I'm not a fan of either option, honestly (plainly written type annotations > turbofishing, but the repetition *is* bothersome); for now I have just taken the systematic option.

**Review by dtolnay — APPROVED** [review]
Thanks!

I am attempting to start a T-libs-api FCP for this in https://github.com/rust-lang/rust/issues/96097#issuecomment-2150973405.

https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/rfcbot.20asleep

**Comment by Amanieu** [comment]
We discussed this in the @rust-lang/libs-api meeting today. We are happy to remove `uninit_array`, but would like to keep it unstable to see if there is demand for a shorter-to-write version of the const block variant.

As such we would like to:
- Point people to the const block syntax in the docs for `uninit_array`.
- If after 2-3 releases of having stabilized const blocks nobody has commented in the tracking issue for `uninit_array` then just remove it.

**Comment by kpreid** [comment]
@Amanieu What should be done with the _usage_ of `uninit_array()`? In other words, should this PR be just closed, or should it be trimmed down to only remove usage and not remove the function?

**Comment by clarfonthey** [comment]
Presumably the answer would be deprecating the method and pointing to the tracking issue to get more feedback?

I still think that the correct answer would be the ability to index into `MaybeUninit<[T; N]>` directly to get `MaybeUninit<T>` references, but every time I bring this up folks don't seem to keen on the idea.

**Comment by dtolnay** [comment]
> What should be done with the _usage_ of `uninit_array()`? In other words, should this PR be just closed, or should it be trimmed down to only remove usage and not remove the function?

I would like to accept all the other changes in this PR other than the deletion of the function. I think we should go ahead and use our own recommended replacement in rustc that we are recommending everyone outside rustc to use.

I would also accept a `#[deprecated = "..."]` attribute on `uninit_array` with guidance in the deprecation message to use `[MaybeUninit::uninit(); N]` or `[const { MaybeUninit::uninit() }; N]` instead.

> I still think that the correct answer would be the ability to index into `MaybeUninit<[T; N]>` directly to get `MaybeUninit<T>` references, but every time I bring this up folks don't seem too keen on the idea.

@clarfonthey do you have a link to this getting shot down? I am on board with this `Index` and `IndexMut` impl being added.

**Comment by clarfonthey** [comment]
I don't have any specific examples of an `Index` impl being shot down, although I can go and take a look later today to find some of the specific discussions I've been thinking of.

Most of them concern specifically impls for `MaybeUninit<[T]>` and the existence of that type in general, which has mostly been met with apprehension. It would require dedicated compiler support to make that work, since unsized unions aren't a thing right now.

Of course, for arrays, that restriction doesn't apply.

**Comment by kpreid** [comment]
> I would like to accept all the other changes in this PR other than the deletion of the function.

Updated.

**Review by dtolnay — APPROVED** [review]

**Pull request merged**
